File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/feature/logging/DevLogger.kt
package io.nativeblocks.devkit.feature.logging

import io.nativeblocks.runtime.api.provider.logger.INativeLogger
import io.nativeblocks.runtime.api.provider.logger.LoggerEventLevel
import io.nativeblocks.devkit.feature.logging.domain.repository.LoggingRepository
import io.nativeblocks.devkit.util.DevKitLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

internal class DevLogger(
    private val loggingRepository: LoggingRepository
) : INativeLogger {

    companion object {
        private val TAG = DevLogger::class.java.simpleName
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    override fun log(
        level: LoggerEventLevel,
        event: String,
        message: String,
        parameters: Map<String, String>
    ) {
        DevKitLogger.d(TAG, "Logging event: $event, level: $level, message: $message")

        scope.launch {
            try {
                loggingRepository.sendLog(
                    level = level,
                    event = event,
                    message = message,
                    parameters = parameters
                )
            } catch (e: Exception) {
                DevKitLogger.e(TAG, "Failed to send log: ${e.message}", e)
            }
        }
    }
}

File: platforms/ios/devkit/Sources/NativeblocksDevkit/feature/logging/DevLogger.swift
import Foundation
import NativeblocksRuntime

internal class DevLogger: INativeLogger {
    private static let TAG = "DevLogger"

    private let loggingRepository: LoggingRepository
    private let queue = DispatchQueue(label: "io.nativeblocks.devkit.devlogger", qos: .utility)

    init(repository: LoggingRepository) {
        self.loggingRepository = repository
    }

    func log(
        level: LoggerEventLevel,
        event: String,
        message: String,
        parameters: [String: String]
    ) {
        DevKitLogger.debug(Self.TAG, "Logging event: \(event), level: \(level.rawValue), message: \(message)")

        queue.async { [weak self] in
            guard let self = self else { return }

            Task {
                do {
                    try await self.loggingRepository.sendLog(
                        level: level,
                        event: event,
                        message: message,
                        parameters: parameters
                    )
                } catch {
                    DevKitLogger.debug(Self.TAG, "Failed to send log: \(error.localizedDescription)")
                }
            }
        }
    }
}
